OSDOCS#6222: Adding custom security groups in AWS

mjpytlak opened the pull request:
Version(s):
4.14+

Issue:
This issue addresses [osdocs-6222](https://issues.redhat.com/browse/OSDOCS-6222).

Link to docs preview:

- [AWS security groups](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-vpc.html#installation-aws-security-groups_installing-aws-vpc)
- Installing a cluster on AWS into an existing VPC > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-vpc.html#installation-aws-vpc-security-groups_installing-aws-vpc)
- Installing a private cluster on AWS > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-private.html#installation-aws-vpc-security-groups_installing-aws-private)
- Installing a cluster on AWS into a government region > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-government-region.html#installation-aws-vpc-security-groups_installing-aws-government-region)
- Installing a cluster on AWS into a Secret or Top Secret Region > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-secret-region.html#installation-aws-vpc-security-groups_installing-aws-secret-region)
- Installing a cluster on AWS China > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-china.html#installation-aws-vpc-security-groups_installing-aws-china-region)
- Installing a cluster using AWS Local Zone > [The edge compute pool for AWS Local Zones](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#machines-edge-machine-pool_installing-aws-localzone)
- Installing a cluster on AWS with remote workers on AWS Outposts > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-outposts-remote-workers.html#installation-aws-vpc-security-groups_installing-aws-outposts-remote-workers)

QE review:
- [ ] QE has approved this change.

mtulio commented on modules/installation-aws-vpc-security-groups.adoc:
```diff
@@ -0,0 +1,55 @@
+      additionalSecurityGroupIDs:
+        - sg-1
+        - sg-2
+  replicas: 3
```
I think it's important to make sure that feature also can work with the edge compute pool with Local Zones.
https://docs.openshift.com/container-platform/4.13/installing/installing_aws/installing-aws-localzone.html

@barbacbd can you see any restriction on that?

barbacbd replied:
@mtulio as far as local zones go, security groups don't require anything special. The security groups just need to be attached to the VPC with the local zone subnet. Also we would require that these local zone subnets are provided in the subnets field of the install config.

barbacbd replied:
@mtulio as far as edge computing, what concerns do you have? Security groups can be shared by many resources as they are just rules.

mtulio replied:
> @mtulio as far as edge computing, what concerns do you have? Security groups can be shared by many resources as they are just rules.

Yes, SGs can be assigned to any supported resource between the same VPC.

My concern was on the install side, I just wanna make sure the configuration below can be used - considering `edge` compute pool uses the same flow of `worker`, I expect only `sg-5` and `sg-6` in MachineSet manifests for Local Zones, just confirming it.

```yaml
compute:
- hyperthreading: Enabled
  name: worker
  platform:
    aws:
      additionalSecurityGroupIDs:
        - sg-1 <1>
        - sg-2
  replicas: 3
- hyperthreading: Enabled
  name: edge
  platform:
    aws:
      additionalSecurityGroupIDs:
        - sg-5
        - sg-6
```

barbacbd replied:
Confirmed! The new machineset does show that it is for edge nodes and it has the security group that I mentioned in the install config and not the one used for the workers

barbacbd commented:
/cc @barbacbd

barbacbd commented on modules/installation-aws-vpc-security-groups.adoc:
```diff
@@ -0,0 +1,55 @@
+platform:
+  aws:
+    region: us-east-1
+    subnets:
```
@mjpytlak did we want to note that the subnets are required or just let the user figure that out during install?

mjpytlak replied:
> did we want to note that the subnets are required or just let the user figure that out during install?

@barbacbd I was originally thinking that given an earlier sample file in the topic states that the subnets are required, I would let it be. However - can't hurt. I updated code block with a few annotations. Outside of that, I made your proposed changes. Thanks for those. Please let me know when you have had a chance to look into the Local Zones case.

mjpytlak commented:
@nirarg I am working with @barbacbd on the ability to apply existing AWS security groups when users BYON. I noticed that the AWS Outposts case requires users to BYON. The content I am updating is shared by the Outposts doc. Do either you or Brent have concerns about this content appear in that installation use case? [1]

[1] https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-outposts-remote-workers.html

mjpytlak commented:
@barbacbd Following up on whether we can include the content on security groups to the Local Zones install use case. Do either you or @nirarg  have any concerns about adding this content to the AWS Outpost install use case?

barbacbd commented:
@mjpytlak @nirarg I don't see any issues with using custom security groups with any BYON. As long as the security groups are attached to a VPC then there should be no issues.

barbacbd commented:
/lgtm

mjpytlak commented:
@mtulio I have added this content to the Local Zones doc. Notably I have updated the sample example to include local zone specific content. PTAL and let me know if we are OK as is or if updates are required. Thanks in advance.

- https://github.com/openshift/openshift-docs/pull/60921/files#diff-c1a8e33036504600b04232682d3adb385bd0cb256c8bb0fa17b50c6daec36635R38
- https://github.com/openshift/openshift-docs/pull/60921/files#diff-c1a8e33036504600b04232682d3adb385bd0cb256c8bb0fa17b50c6daec36635R65

mjpytlak commented on modules/installation-applying-aws-security-groups.adoc:
```diff
@@ -0,0 +1,86 @@
+  name: worker
+endif::local-zones[]
+ifdef::local-zones[]
+  - name: edge
```
I am interpreting your comment as `ifdef::local-zones[ ]` is incorrect. Are you referring to another hyphen in this example?

mtulio replied:
@mjpytlak - looking to the rendering page from this PR:
https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#installation-aws-vpc-security-groups_installing-aws-localzone

![Screenshot from 2023-07-13 16-13-51](https://github.com/openshift/openshift-docs/assets/3216894/0bfe25b8-3783-4bfe-a25f-b35fcad18657)

mtulio replied:
@mjpytlak Also looking at where that section is placed, I can see the docs is providing different examples of optional machine configuration (like Security Groups) here:
https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#machines-edge-machine-pool_installing-aws-localzone

- Configuration that uses an edge pool with default settings
- Configuration that uses an edge pool with a custom instance type
- Configuration that uses an edge pool with a custom EBS type

Do you think that optional security group deployment would fit in that list too, as it is also optional?

mtulio replied:
@mjpytlak ok, thanks,  I can see there. Do we need two examples about that topic on the Local Zone page? As you agreed to add the example[1] in the section, I can't see a benefit of [2] (as we already have detailed in the other pages). As a user [reading the rendered page] it seems the SG section ([2][Applying existing AWS security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#installation-aws-vpc-security-groups_installing-aws-localzone)) is required in the installation process for Local Zones - which is not true.

- (1) Installing a cluster using AWS Local Zone > [The edge compute pool for AWS Local Zones](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#machines-edge-machine-pool_installing-aws-localzone)
- (2) Installing a cluster using AWS Local Zone > [Applying existing security groups to a cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#installation-aws-vpc-security-groups_installing-aws-localzone)

I am ok having only 1 on the Local Zones install page. Let me know your thoughts.

mjpytlak replied:
> I am ok having only 1 on the Local Zones install page. Let me know your thoughts.

@mtulio I agree and that was an oversight on my part. I removed Applying existing AWS security groups to the cluster.

mtulio replied:
@mjpytlak np at all! Thanks for revisiting.

mtulio commented on modules/installation-applying-aws-security-groups.adoc:
```diff
@@ -0,0 +1,86 @@
+[source,yaml]
+----
+#...
+compute:
+- hyperthreading: Enabled
+ifndef::local-zones[]
+  name: worker
+endif::local-zones[]
+ifdef::local-zones[]
+  - name: edge
```
```suggestion
compute:
- hyperthreading: Enabled
ifndef::local-zones[]
  name: worker
endif::local-zones[]
ifdef::local-zones[]
  name: edge
```

mjpytlak replied:
Done. Thank you for the clarification on Friday. I did not know the `-` from `hyperthreading` was inherited.

mtulio replied:
lgtm that part. Thanks

mjpytlak commented:
@yunjiang29 Draft docs ready for QE review. PTAL. Thank you.

yunjiang29 commented:
@gpei PTAL, thanks

gpei commented:
/lgtm

bergerhoffer commented:
/label peer-review-in-progress

bergerhoffer commented on modules/installation-applying-aws-security-groups.adoc:
```diff
@@ -0,0 +1,58 @@
+
+.Procedure
+
+. Edit the `compute.platform.aws.additionalSecurityGroupIDs` parameter to specify one or more custom security groups for your compute machines.
```
Where are we editing, in the install-config.yaml file? I'd clarify that.

bergerhoffer commented on modules/installation-applying-aws-security-groups.adoc:
```diff
@@ -0,0 +1,58 @@
+.Procedure
+
+. Edit the `compute.platform.aws.additionalSecurityGroupIDs` parameter to specify one or more custom security groups for your compute machines.
+. Edit `controlPlane.platform.aws.additionalSecurityGroupIDs` parameter to specify one or more custom security groups for your control plane machines.
```
```suggestion
. Edit the `controlPlane.platform.aws.additionalSecurityGroupIDs` parameter to specify one or more custom security groups for your control plane machines.
```

bergerhoffer commented on modules/installation-applying-aws-security-groups.adoc:
```diff
@@ -0,0 +1,58 @@
+. Edit `controlPlane.platform.aws.additionalSecurityGroupIDs` parameter to specify one or more custom security groups for your control plane machines.
+. Save the file and reference it when deploying the cluster.
+
+.Sample `install-config.yaml` files that specifies custom security groups
```
```suggestion
.Sample `install-config.yaml` file that specifies custom security groups
```

bergerhoffer commented on modules/installation-applying-aws-security-groups.adoc:
```diff
@@ -0,0 +1,58 @@
+.Sample `install-config.yaml` files that specifies custom security groups
+[source,yaml]
+----
+#...
```
There should be a space, `# ...`

bergerhoffer commented on modules/installation-aws-security-groups.adoc:
```diff
@@ -0,0 +1,34 @@
+
+By default, the installation program creates and attaches security groups to control plane and compute machines. The rules associated with the default security groups cannot be modified.
+
+However, you can apply additional existing AWS security groups (custom security groups), which are associated with your existing VPC, to control plane and compute machines. Applying custom security groups can help you meet the security needs of your organization, in such cases where you need to control the incoming or outgoing traffic of these machines.
```
Not critical, but I'd argue that you don't need the `(custom security groups)`

bergerhoffer reviewed (commented):
Nice job, just a few suggestions

bergerhoffer commented:
/label peer-review-done
/remove-label peer-review-needed
/remove-label peer-review-in-progress

mjpytlak commented:
Appreciate the review @bergerhoffer. Always helpful.

mjpytlak commented:
/cherrypick enterprise-4.14
